File: app/src/main/java/com/pro/app/data/network/NetworkConnectionInterceptor.java
package com.pro.app.data.network;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkRequest;
import android.os.Build;
import com.appizona.yehiahd.fastsave.FastSave;
import com.mindorks.nybus.NYBus;
import com.pro.app.MainApplication;
import com.pro.app.data.events.NoInternetEvent;
import com.pro.app.utils.Constants;

import java.io.IOException;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

public abstract class NetworkConnectionInterceptor implements Interceptor {

    @Override
    public Response intercept(Chain chain) throws IOException {

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            if (!MainApplication.Companion.getInstance().isInternetAvailable()) {
                throw new NoConnectivityException();
            }
        }
        else
        {
            ConnectivityManager connectivityManager = (ConnectivityManager) MainApplication.Companion.getInstance().getApplicationContext().getSystemService(Context.CONNECTIVITY_SERVICE);

            NetworkRequest.Builder builder = new NetworkRequest.Builder();

            connectivityManager.registerNetworkCallback(
                    builder.build(),
                    new ConnectivityManager.NetworkCallback() {
                        @Override
                        public void onAvailable(Network network) {
                        }
                        @Override
                        public void onLost(Network network) {
                            NYBus.get().post(new NoInternetEvent());
                        }
                    }

            );
        }

        Request original = chain.request();
        String android_id = MainApplication.Companion.getInstance().getDeviceID();

        String token = FastSave.getInstance().getString(Constants.SP_USER_TOKEN, "");
        token = (token != null) ? token : "";

        Request request = original.newBuilder()
                .header("token", token)
                .header("os", "android")
                .header("duid", android_id)
                .header("version", "1")
                .header("app_version", BuildConfig.VERSION_NAME)
                .header("app_version_code", BuildConfig.VERSION_CODE + "")
                .method(original.method(), original.body())
                .build();
        return chain.proceed(request);
    }

    class NoConnectivityException extends IOException {

        @Override
        public String getMessage() {
            return "Network Connection exception";
        }

    }
}

File: app/src/main/java/com/pro/app/data/network/NetworkConnectionInterceptor.kt
package com.pro.app.data.network

import android.content.Context
import android.net.ConnectivityManager
import android.net.ConnectivityManager.NetworkCallback
import android.net.Network
import android.net.NetworkRequest
import android.os.Build
import com.mindorks.nybus.NYBus
import com.pro.app.BuildConfig
import com.pro.app.MainApplication.Companion.instance
import com.pro.app.data.events.NoInternetEvent
import com.pro.app.utils.Constants
import okhttp3.Interceptor
import okhttp3.Response
import java.io.IOException

abstract class NetworkConnectionInterceptor : Interceptor {

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            if (!instance.isInternetAvailable()) {
                throw NoConnectivityException()
            }
        } else {
            val connectivityManager =
                instance.applicationContext
                    .getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            val builder = NetworkRequest.Builder()
            connectivityManager.registerNetworkCallback(
                builder.build(),
                object : NetworkCallback() {
                    override fun onAvailable(network: Network) {}
                    override fun onLost(network: Network) {
                        NYBus.get().post(NoInternetEvent())
                    }
                }
            )
        }
        val original = chain.request()
        val android_id = instance.deviceID

        val request = original.newBuilder()
            .header("Authorization", "Bearer eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiI4MTFiYTE2YzFlN2VkNzRlNmI3YTA1YzRmZmJmY2UzOSIsInN1YiI6IjVmOTFlMTQwZDlmNGE2MDAzN2Q5ZDg2YiIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.eKtfOeKlSsrI2R5LmoOCnZLQ3R0g2kA7n4lbX4kFcTc")
            .header("os", "android")
            .header("duid", android_id)
            .header("app_version", BuildConfig.VERSION_NAME)
            .header("app_version_code", BuildConfig.VERSION_CODE.toString())
            .method(original.method(), original.body())
            .build()
        return chain.proceed(request)
    }

    internal inner class NoConnectivityException : IOException() {
        override val message: String
            get() = "Network Connection exception"
    }
}